### PR TITLE
Allow multiple destinations in egress-router

### PR DESCRIPTION
**Pull request by danwinship**
The current egress-router maps from any source port to the corresponding port on a single destination IP. A bunch of customers have requested the ability to have multiple destination IPs, by mapping different source ports to different destination IP. This implements that (along with some general egress-router cleanups, and preparation for the revised version of #13586).

The use of multi-line environment variables is a little bit awkward, but yaml actually makes it pretty easy... Also, we plan to update the docs to explain how to use a ConfigMap rather than including the environment variable values literally in the Pod spec, which will make it a bit nicer. (This is also why the syntax allows blank lines and comments, which may be useful if you're generating a ConfigMap from a file.)

Is this trying to do too much with bash? In particular, if you use a ConfigMap to set the environment variables, then you'll have a privileged pod trying to parse complicated environment variables set by a user who may not have had permission to create a privileged pod themselves, which creates the possibility of hacking. (If a user hacks into the egress-router pod, they can, at a minimum, sniff all inter-node pod-to-pod traffic, and possibly spoof it as well.) Eg, the current version of egress-router doesn't quote things at all and would be hackable via something like `EGRESS_SOURCE="]; then echo HACKED; fi; if ["`. The updated version tries to be a lot better about quoting, and verifying that data has the expected form, but it's hard to feel completely confident...

Trello card: https://trello.com/c/VLnnrZLG

@openshift/networking PTAL
@stevekuznetsov, you may have opinions on the shell scripting stuff

### CODE REVIEW

File: images/egress/router/egress-router.sh [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+if [[ -z "${EGRESS_DESTINATION:-}" ]]; then
+    echo "EGRESS_DESTINATION unspecified"
+    exit 1
+fi
+if [[ "${EGRESS_GATEWAY:-}" !~ ^${IP_REGEX}$ ]]; then
+    echo "EGRESS_GATEWAY unspecified or invalid"
+    exit 1
+fi
+
+set -x
```
Comment (stevekuznetsov) [inline images/egress/router/egress-router.sh]: Left from debugging?

File: images/egress/router/egress-router.sh [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    ip route add "${EGRESS_GATEWAY}"/32 dev macvlan0
+    ip route del default
+    ip route add default via "${EGRESS_GATEWAY}" dev macvlan0
+fi
+
+# Update neighbor ARP caches in case another node previously had the IP. (This is
+# the same code ifup uses.)
+arping -q -A -c 1 -I macvlan0 "${EGRESS_SOURCE}"
+( sleep 2;
+  arping -q -U -c 1 -I macvlan0 "${EGRESS_SOURCE}" || true ) > /dev/null 2>&1 < /dev/null &
```
Comment (stevekuznetsov) [inline images/egress/router/egress-router.sh]: Why doesn't it matter if this fails?

File: images/egress/router/egress-router.sh (hunk 1 above)
Comment (danwinship) [inline images/egress/router/egress-router.sh]: Yes and no; it was there before, just earlier in the script. But I guess it's only useful for debugging purposes; if an actual error occurs then it should log something that will show up in "oc logs" anyway. So I'll make it not do that by default.

File: images/egress/router/egress-router.sh (hunk 2 above)
Comment (danwinship) [inline images/egress/router/egress-router.sh]: Pretty sure I just copied the "|| true" from ifup-eth, although the current version doesn't have that... I'll look into why arping might fail and reconsider if want it

**Comment by stevekuznetsov** [comment]
> Is this trying to do too much with bash?

I struggle with answering this question with anything other than "yes", by the way. It's a lot of code and a lot of logic, but there's no way to unit-test it and no way to ensure it stays functional through CI other than at very high levels of integration tests. I think the development and maintenance burden of a Bash script over a Python script (or whatever language that makes it simple to shell out) often is hard to justify. 

That being said, I'm more guilty than anyone for not following that advice when momentum already has Bash as the driving force for a project, and it becomes especially difficult to follow when there's a lot of shelling out to interact with tools that only have a CLI.

**Comment by danwinship** [comment]
> Do we make sure the IP is not in a cluster managed CIDR?

We don't. I think it doesn't work if you try, though I don't remember for sure...

**Comment by danwinship** [comment]
> there's no way to unit-test it

hm... says who?

New version adds a unit test (written in go to integrate with the rest of our unit testing) that runs egress-router.sh in a special unit-testing mode that just outputs the iptables rules it would set. Seems to work...

[test]

**Comment by stevekuznetsov** [comment]
I think that defines the "unit" at a little bit of a larger scale than usual "units". I didn't say it was untestable :)

File: images/egress/router/egress-router.sh [hunk 3]
```diff
@@ -1,38 +1,44 @@
...
 fi
 
+set -x
+
 # The pod may die and get restarted; only try to add the
 # address/route/rules if they are not already there.
-if ! ip route get ${EGRESS_DESTINATION} | grep -q macvlan0; then
-    ip addr add ${EGRESS_SOURCE}/32 dev macvlan0
+if ! ip route get "${EGRESS_DESTINATION}" | grep -q macvlan0; then
+    ip addr add "${EGRESS_SOURCE}"/32 dev macvlan0
```
Comment (dcbw) [inline images/egress/router/egress-router.sh]: That's a wierd syntax, but /sbin/ip accepts it without question, so...

Comment (danwinship) [inline images/egress/router/egress-router.sh]: what's weird about it?

File: images/egress/router/egress-router.sh [hunk 4]
```diff
@@ -0,0 +1,136 @@
+#!/bin/bash
+
+# OpenShift egress router setup script
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+BLANK_LINE_OR_COMMENT_REGEX="([[:space:]]*$|#.*)"
+PORT_REGEX="[[:digit:]]+"
```
Comment (danwinship) [inline images/egress/router/egress-router.sh]: iptables-restore will fail if the data we pass it is bad, so we don't need to validate it 100% here. The regexes are about (a) distinguishing the different valid inputs (which is more important in the http proxy case (#13586) where we have to distinguish "a line which is just a hostname" from "a line which is just an IP address") and (b) making sure that when we plug them into an iptables rule, iptables will either parse it correctly or reject it entirely. (Eg, we don't want to allow `destip` to contain spaces and random iptables flags.)

But I can adjust the strictness of the checks here if people want.

**Comment by knobunc** [comment]
[merge]

**Comment by danwinship** [comment]
flake... maybe infrastructure? #13944 is similar and was also considered to be infrastructure...
[merge]

**Pull request merged**